### PR TITLE
fix txpool underpriced check

### PR DESCRIPTION
**Pull request by 0xmountaintop**
## 1. Purpose or design rationale of this PR

https://github.com/scroll-tech/go-ethereum/pull/825 + https://github.com/scroll-tech/go-ethereum/pull/901

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Review by colinlyguo — DISMISSED** [review]

File: core/txpool/legacypool/legacypool.go [hunk 1]
```diff
@@ -746,7 +746,8 @@ func (pool *LegacyPool) add(tx *types.Transaction, local bool) (replaced bool, e
 	// If the transaction pool is full, discard underpriced transactions
 	if uint64(pool.all.Slots()+numSlots(tx)) > pool.config.GlobalSlots+pool.config.GlobalQueue {
 		// If the new transaction is underpriced, don't accept it
-		if !isLocal && pool.priced.Underpriced(tx) {
+		// if !isLocal && pool.priced.Underpriced(tx) {
+		if !isLocal && tx.GasFeeCapIntCmp(pool.gasTip.Load()) < 0 {
 			log.Trace("Discarding underpriced transaction", "hash", hash, "gasTipCap", tx.GasTipCap(), "gasFeeCap", tx.GasFeeCap())
```
Comment (colinlyguo) [inline core/txpool/legacypool/legacypool.go]:
```suggestion
			log.Trace("Discarding underpriced transaction", "hash", hash, "gasTipCap", tx.GasTipCap(), "gasFeeCap", tx.GasFeeCap(), "txPoolGasTip", pool.gasTip.Load())
```

File: core/txpool/legacypool/legacypool.go [hunk 2]
```diff
@@ -748,7 +748,7 @@ func (pool *LegacyPool) add(tx *types.Transaction, local bool) (replaced bool, e
 		// If the new transaction is underpriced, don't accept it
 		// if !isLocal && pool.priced.Underpriced(tx) {
 		if !isLocal && tx.GasFeeCapIntCmp(pool.gasTip.Load()) < 0 {
-			log.Trace("Discarding underpriced transaction", "hash", hash, "gasTipCap", tx.GasTipCap(), "gasFeeCap", tx.GasFeeCap())
+			log.Trace("Discarding underpriced transaction", "hash", hash, "gasTipCap", tx.GasTipCap(), "gasFeeCap", tx.GasFeeCap(), "txPoolGasTip", pool.gasTip.Load())
```
Comment (colinlyguo) [inline core/txpool/legacypool/legacypool.go:751]: sorry that I found a potential data inconsistency that is introduced by my comment... if `pool.gasTip.Store()` is called in between (after the `GasFeeCapIntCmp` check and before printing the `Trace` log), `pool.gasTip.Load()` would be different.

and `pool.gasTip.Store()` is only called by `SetGasTip` after the legacy pool is initialized. What do you think, @HAOYUatHZ? I'm OK with any refactors about this.

**Review by colinlyguo — APPROVED** [review]

File: core/txpool/legacypool/legacypool.go (hunk 2 above)
Comment (0xmountaintop) [reply core/txpool/legacypool/legacypool.go:751]: I think we can keep it as what it is now

**Review by colinlyguo — APPROVED** [review]

File: core/txpool/legacypool/legacypool.go [hunk 3]
```diff
@@ -746,8 +746,9 @@ func (pool *LegacyPool) add(tx *types.Transaction, local bool) (replaced bool, e
 	// If the transaction pool is full, discard underpriced transactions
 	if uint64(pool.all.Slots()+numSlots(tx)) > pool.config.GlobalSlots+pool.config.GlobalQueue {
 		// If the new transaction is underpriced, don't accept it
-		if !isLocal && pool.priced.Underpriced(tx) {
-			log.Trace("Discarding underpriced transaction", "hash", hash, "gasTipCap", tx.GasTipCap(), "gasFeeCap", tx.GasFeeCap())
+		// if !isLocal && pool.priced.Underpriced(tx) {
+		if !isLocal && tx.GasFeeCapIntCmp(pool.gasTip.Load()) < 0 {
```
Comment (Thegaram) [inline core/txpool/legacypool/legacypool.go:750]: The name `gasTip` is a bit counter-intuitive here. But this impl looks consistent with our current node, so LGTM.

**Review by Thegaram — APPROVED** [review]

**Pull request merged**